nemo/collections/asr/parts/utils/online_clustering.py

```diff
@@ -1189,3 +1196,422 @@ def forward_infer(
+            Y_part = self.forward_unit_infer(
+                emb_part,
+                oracle_num_speakers=self.sub_cluster_n,
+                max_rp_threshold=max_rp_threshold,
+                max_num_speakers=self.sub_cluster_n,
+                sparse_search_volume=sparse_search_volume,
+                cuda=self.cuda,
+            )
+            num_to_be_merged = int(min(unit_window_len, emb_part.shape[0]) - sub_cluster_n)
+            min_count_per_cluster = self.get_div_ceil_count(
+                numer=self.sub_cluster_n, denomin=len(torch.unique(Y_part))
+            )
+
+            # We want only one embedding vector for each cluster, so we calculate the number of embedding vectors to be removed
+            class_target_vol = get_merge_quantity(
+                num_to_be_removed=num_to_be_merged,
+                pre_clus_labels=Y_part,
+                min_count_per_cluster=min_count_per_cluster,
+            )
+            if not torch.jit.is_scripting():
+                pbar.update(1)
+
+            # `class_target_vol` is a list of cluster-indices from overclustering
+            for spk_idx, merge_quantity in enumerate(list(class_target_vol)):
+                merged_embs, merged_clus_labels, index_mapping = run_reducer(
+                    pre_embs=emb_part, target_spk_idx=spk_idx, merge_quantity=merge_quantity, pre_clus_labels=Y_part,
+                )
+                total_emb.append(merged_embs)
+                absolute_index_mapping = [x + offset_index for x in index_mapping]
+                absolute_merge_mapping.append(absolute_index_mapping)
+                window_range_list.append([window_offset, window_offset + merged_embs.shape[0]])
+                window_offset += merged_embs.shape[0]
+
+        if not torch.jit.is_scripting():
+            pbar.close()
+
+        # Concatenate the reduced embeddings then perform high-level clustering
+        reduced_embs = torch.cat(total_emb)
+        Y_aggr = self.forward_unit_infer(
+            reduced_embs,
+            oracle_num_speakers=oracle_num_speakers,
+            max_rp_threshold=max_rp_threshold,
+            max_num_speakers=max_num_speakers,
+            sparse_search_volume=sparse_search_volume,
+            fixed_thres=fixed_thres,
+            cuda=self.cuda,
+        )
+
+        Y_unpack = self.unpack_labels(
+            Y_aggr=Y_aggr,
+            window_range_list=window_range_list,
+            absolute_merge_mapping=absolute_merge_mapping,
+            org_len=emb.shape[0],
+        )
+        if reduced_embs.shape[0] != Y_aggr.shape[0]:
+            raise ValueError(
+                "The number of embeddings (reduced_embs.shape) and the number of labels (Y_aggr.shape) do not match."
+            )
+        return Y_unpack
+
+    def forward_infer(
+        self,
+        embeddings_in_scales: torch.Tensor,
+        timestamps_in_scales: torch.Tensor,
```

Comment:
is it `List[torch.Tensor]` for these two args?

Reply:
It is passed as concatenated torch Tensor so this form is correct.
If Type hint is wrong, torch jit compiler breaks at this line.